File: ImapX.Sample/FolderBox.cs
using System;
using System.Linq;
using System.Windows.Forms;

namespace ImapX.Sample
{
    public sealed partial class FolderBox : Form
    {
        public FolderBox(string title, string text)
        {
            InitializeComponent();

            Text = title;
            lblTitle.Text = title;
            lblText.Text = text;

            trwFolders.Nodes.Add(Program.ImapClient.Host);
            trwFolders.Nodes[0].Nodes.AddRange(Program.ImapClient.Folders.Select(FolderToNode).ToArray());
            trwFolders.Nodes[0].Expand();
        }

        public Folder SelectedFolder
        {
            get { return trwFolders.SelectedNode.Tag as Folder; }
        }

        private TreeNode FolderToNode(Folder folder)
        {
            var node = new TreeNode(folder.Name);
            node.Nodes.AddRange(folder.SubFolders.Select(FolderToNode).ToArray());
            node.Tag = folder;
            return node;
        }

        private void btnOK_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.OK;
            Close();
        }

        private void trwFolders_AfterSelect(object sender, TreeViewEventArgs e)
        {
            btnOK.Enabled = trwFolders.SelectedNode != null && trwFolders.SelectedNode != trwFolders.Nodes[0];
        }

        private void trwFolders_BeforeSelect(object sender, TreeViewCancelEventArgs e)
        {
            var folder = e.Node.Tag as Folder;
            if (folder == null || !folder.Selectable)
                e.Cancel = true;
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
            Close();
        }

        public static Folder Show(string title, string text, IWin32Window owner = null)
        {
            using (var dlg = new FolderBox(title, text))
                if (dlg.ShowDialog(owner) == DialogResult.OK)
                    return dlg.SelectedFolder;
            return null;
        }

        private void trwFolders_NodeMouseDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
        {
            if (btnOK.Enabled)
                btnOK_Click(null, null);
        }
    }
}

File: ImapX.VBSample/FolderBox.vb
Public Class FolderBox

    Public Sub New(ByVal title As String, ByVal [text] As String)
        Me.InitializeComponent()
        Me.Text = title
        Me.lblTitle.Text = title
        Me.lblText.Text = [text]
        Me.trwFolders.Nodes.Add(My.MyApplication.ImapClient.Host)
        Me.trwFolders.Nodes.Item(0).Nodes.AddRange(Enumerable.ToArray(Of TreeNode)(Enumerable.Select(Of Folder, TreeNode)(My.MyApplication.ImapClient.Folders, New Func(Of Folder, TreeNode)(AddressOf Me.FolderToNode))))
        Me.trwFolders.Nodes.Item(0).Expand()
    End Sub

    Private Sub btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs)
        MyBase.DialogResult = DialogResult.Cancel
        MyBase.Close()
    End Sub

    Private Sub btnOK_Click(ByVal sender As Object, ByVal e As EventArgs)
        MyBase.DialogResult = DialogResult.OK
        MyBase.Close()
    End Sub

    Private Function FolderToNode(ByVal folder As Folder) As TreeNode
        Dim node As New TreeNode(folder.Name)
        node.Nodes.AddRange(Enumerable.ToArray(Of TreeNode)(Enumerable.Select(Of Folder, TreeNode)(folder.SubFolders, New Func(Of Folder, TreeNode)(AddressOf Me.FolderToNode))))
        node.Tag = folder
        Return node
    End Function

    Public Overloads Shared Function Show(ByVal title As String, ByVal [text] As String, Optional ByVal owner As IWin32Window = Nothing) As Folder
        Using dlg As FolderBox = New FolderBox(title, [text])
            If (dlg.ShowDialog(owner) = DialogResult.OK) Then
                Return dlg.SelectedFolder
            End If
        End Using
        Return Nothing
    End Function

    Private Sub trwFolders_AfterSelect(ByVal sender As Object, ByVal e As TreeViewEventArgs)
        Me.btnOK.Enabled = ((Not Me.trwFolders.SelectedNode Is Nothing) AndAlso (Not Me.trwFolders.SelectedNode Is Me.trwFolders.Nodes.Item(0)))
    End Sub

    Private Sub trwFolders_BeforeSelect(ByVal sender As Object, ByVal e As TreeViewCancelEventArgs)
        Dim folder As Folder = TryCast(e.Node.Tag, Folder)
        If Not ((Not folder Is Nothing) AndAlso folder.Selectable) Then
            e.Cancel = True
        End If
    End Sub

    Private Sub trwFolders_NodeMouseDoubleClick(ByVal sender As Object, ByVal e As TreeNodeMouseClickEventArgs)
        If Me.btnOK.Enabled Then
            Me.btnOK_Click(Nothing, Nothing)
        End If
    End Sub


    ' Properties
    Public ReadOnly Property SelectedFolder As Folder
        Get
            Return TryCast(Me.trwFolders.SelectedNode.Tag, Folder)
        End Get
    End Property


End Class